float32[] angles
float32[] ranges
float32[] type
